File: src/main/java/io/github/chaosunity/forgelin/preloader/ForgelinSetup.java
package io.github.chaosunity.forgelin.preloader;

import net.minecraftforge.fml.relauncher.IFMLCallHook;

import java.util.Map;

public class ForgelinSetup implements IFMLCallHook {
    @Override
    public void injectData(Map<String, Object> data) {
        ClassLoader loader = (ClassLoader) data.get("classLoader");
        try {
            loader.loadClass("io.github.chaosunity.forgelin.KotlinAdapter");
        } catch (ClassNotFoundException e) {
            // this should never happen
            throw new RuntimeException("Couldn't find Forgelin langague adapter, this shouldn't be happening", e);
        }
    }

    @Override
    public Void call() throws Exception {
        return null;
    }
}

File: src/main/kotlin/io/github/chaosunity/forgelin/preloader/ForgelinSetup.kt
package io.github.chaosunity.forgelin.preloader

import net.minecraftforge.fml.relauncher.IFMLCallHook

@Suppress("unused")
class ForgelinSetup: IFMLCallHook {
    override fun injectData(data: MutableMap<String, Any>?) {
        val loader = data!!["classLoader"]!! as ClassLoader
        
        // Explicit class load for KotlinAdapter
        try {
            loader.loadClass("kotlin.text.Charsets")
            loader.loadClass("io.github.chaosunity.forgelin.KotlinAdapter")
        } catch (e: ClassNotFoundException) {
            // this should never happen
            throw RuntimeException("Couldn't find Forgelin langague adapter, this shouldn't be happening", e)
        }
    }
    
    override fun call(): Void? {
        return null
    }
}